File: ios/Runner/AckQueueStore.swift
import Foundation

/// The same queue Dart uses, written from the native side.
///
/// The ACK action on a notification runs with `foreground: false`, so it can
/// land with no Flutter engine up. The entry goes on this list, which is the
/// exact shape `AckQueue` in Dart reads, and the next launch (or the running
/// app, which is told about it) sends it.
///
/// Keep the field names in step with `lib/core/ack/ack_queue_entry.dart`.
enum AckQueueStore {
    /// `AckQueue.storageKey` in Dart, with the shared_preferences prefix.
    static let key = "flutter.ack_queue_v1"

    /// The same two seconds the Dart queue waits after one failure.
    static let firstRetryMs = 2_000

    static func enqueue(
        action: String,
        incidentId: String,
        alarmFiredAtMs: Int? = nil,
        defaults: UserDefaults = .standard
    ) {
        let now = Int(Date().timeIntervalSince1970 * 1000)
        var entry: [String: Any] = [
            "id": "\(now)-native-\(incidentId)-\(action)",
            "action": action,
            "incident_id": incidentId,
            "enqueued_at_ms": now,
            "attempts": 1,
            "next_attempt_at_ms": now + firstRetryMs,
        ]
        if let alarmFiredAtMs { entry["alarm_fired_at_ms"] = alarmFiredAtMs }

        var entries = read(from: defaults)
        entries.append(entry)
        write(entries, to: defaults)
        NSLog("CritAlarmAck: ack_queued_native action=%@ incident_id=%@", action, incidentId)
    }

    /// Takes every entry for `incidentId` with this `action` off the queue.
    /// For after `NativeAckSender` has already landed it, so Dart does not
    /// send it a second time.
    static func remove(
        incidentId: String,
        action: String,
        defaults: UserDefaults = .standard
    ) {
        let entries = read(from: defaults)
        let kept = entries.filter {
            !($0["incident_id"] as? String == incidentId && $0["action"] as? String == action)
        }
        guard kept.count != entries.count else { return }
        write(kept, to: defaults)
    }

    static func pendingCount(defaults: UserDefaults = .standard) -> Int {
        read(from: defaults).count
    }

    static func debugEntries(defaults: UserDefaults = .standard) -> [[String: Any]] {
        read(from: defaults).compactMap { entry in
            guard let action = entry["action"] as? String,
                  let incidentId = entry["incident_id"] as? String,
                  let attempts = entry["attempts"] as? Int,
                  let nextMs = entry["next_attempt_at_ms"] as? Int else { return nil }
            var row: [String: Any] = [
                "action": action,
                "incident_id": incidentId,
                "attempts": attempts,
                "next_attempt_at": nextMs / 1_000,
            ]
            if let error = entry["last_error"] as? String { row["last_error"] = error }
            return row
        }
    }

    private static func read(from defaults: UserDefaults) -> [[String: Any]] {
        guard let raw = defaults.string(forKey: key),
              let data = raw.data(using: .utf8),
              let decoded = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
        else { return [] }
        return decoded
    }

    /// An empty list is written as no key at all, the way Dart's `_write` does.
    private static func write(_ entries: [[String: Any]], to defaults: UserDefaults) {
        if entries.isEmpty {
            defaults.removeObject(forKey: key)
            return
        }
        guard let data = try? JSONSerialization.data(withJSONObject: entries),
              let json = String(data: data, encoding: .utf8)
        else { return }
        defaults.set(json, forKey: key)
    }
}

File: android/app/src/main/kotlin/app/critalarm/storage/AckQueueStore.kt
package app.critalarm.storage

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * The same queue Dart uses, written from the native side.
 *
 * An ack tapped from a notification is handled here, with no Flutter engine
 * running. When the send fails the entry goes on this list, which is the exact
 * shape `AckQueue` in Dart reads, so the next app launch retries it.
 *
 * Keep the field names in step with `lib/core/ack/ack_queue_entry.dart`.
 */
class AckQueueStore(context: Context) {
    private val preferences = context.getSharedPreferences("FlutterSharedPreferences", Context.MODE_PRIVATE)

    fun enqueue(action: String, incidentId: String, alarmFiredAtMs: Long? = null) {
        val now = System.currentTimeMillis()
        val entry = JSONObject().apply {
            put("id", "$now-native-$incidentId-$action")
            put("action", action)
            put("incident_id", incidentId)
            put("enqueued_at_ms", now)
            put("attempts", 1)
            // Wait the same two seconds the Dart queue waits after one failure.
            put("next_attempt_at_ms", now + FIRST_RETRY_MS)
            if (alarmFiredAtMs != null) put("alarm_fired_at_ms", alarmFiredAtMs)
        }
        val existing = runCatching { JSONArray(preferences.getString(KEY, "[]")) }.getOrNull() ?: JSONArray()
        existing.put(entry)
        preferences.edit().putString(KEY, existing.toString()).apply()
    }

    fun pendingCount(): Int =
        runCatching { JSONArray(preferences.getString(KEY, "[]")).length() }.getOrDefault(0)

    internal fun debugEntries(): List<Map<String, Any>> {
        val entries = runCatching { JSONArray(preferences.getString(KEY, "[]")) }.getOrNull() ?: return emptyList()
        return (0 until entries.length()).mapNotNull { index ->
            val entry = entries.optJSONObject(index) ?: return@mapNotNull null
            val action = entry.optString("action").takeIf(String::isNotEmpty) ?: return@mapNotNull null
            val incidentId = entry.optString("incident_id").takeIf(String::isNotEmpty) ?: return@mapNotNull null
            val attempts = entry.optInt("attempts", -1).takeIf { it >= 0 } ?: return@mapNotNull null
            val next = entry.optLong("next_attempt_at_ms", 0L).takeIf { it > 0 } ?: return@mapNotNull null
            buildMap {
                put("action", action)
                put("incident_id", incidentId)
                put("attempts", attempts)
                put("next_attempt_at", (next / 1_000L).toInt())
                entry.optString("last_error").takeIf(String::isNotEmpty)?.let { put("last_error", it) }
            }
        }
    }

    companion object {
        /** `AckQueue.storageKey` in Dart, with the shared_preferences prefix. */
        const val KEY = "flutter.ack_queue_v1"
        const val FIRST_RETRY_MS = 2_000L
    }
}
